File: backend/src/utils/fechas.js
const config = require('../config');

/** Fecha de hoy 'YYYY-MM-DD' en la zona horaria del negocio (por defecto Lima). */
function hoy(zona = config.zonaHoraria) {
  // en-CA formatea como YYYY-MM-DD
  return new Intl.DateTimeFormat('en-CA', {
    timeZone: zona,
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
  }).format(new Date());
}

/** true si el texto es una fecha real con formato YYYY-MM-DD. */
function esFechaValida(texto) {
  if (typeof texto !== 'string' || !/^\d{4}-\d{2}-\d{2}$/.test(texto)) return false;
  const [a, m, d] = texto.split('-').map(Number);
  const fecha = new Date(Date.UTC(a, m - 1, d));
  return fecha.getUTCFullYear() === a && fecha.getUTCMonth() === m - 1 && fecha.getUTCDate() === d;
}

/** Rangos inclusivos [inicio, fin] que comparten al menos un día. */
function rangosSeSuperponen(a, b) {
  return a.fechaInicio <= b.fechaFin && b.fechaInicio <= a.fechaFin;
}

/** Devuelve el primer par de rangos de la lista que se superponen, o null. */
function buscarSuperposicion(rangos) {
  const ordenados = [...rangos].sort((x, y) => x.fechaInicio.localeCompare(y.fechaInicio));
  for (let i = 1; i < ordenados.length; i++) {
    if (rangosSeSuperponen(ordenados[i - 1], ordenados[i])) {
      return [ordenados[i - 1], ordenados[i]];
    }
  }
  return null;
}

/** Cantidad de días de un rango inclusivo (2026-10-01 a 2026-10-01 = 1 día). */
function diasEntre(fechaInicio, fechaFin) {
  const ms = Date.parse(`${fechaFin}T00:00:00Z`) - Date.parse(`${fechaInicio}T00:00:00Z`);
  return Math.round(ms / 86400000) + 1;
}

module.exports = { hoy, esFechaValida, rangosSeSuperponen, buscarSuperposicion, diasEntre };

File: backend/src/main/java/alquiler/util/Fechas.java
package alquiler.util;

import java.time.LocalDate;
import java.time.ZoneId;
import java.time.format.DateTimeParseException;
import java.time.temporal.ChronoUnit;
import java.util.ArrayList;
import java.util.Comparator;
import java.util.List;
import java.util.regex.Pattern;

/** Utilidades de fechas 'AAAA-MM-DD'. Los rangos son inclusivos: [inicio, fin]. */
public final class Fechas {

    private static final Pattern FORMATO = Pattern.compile("^\\d{4}-\\d{2}-\\d{2}$");

    private Fechas() {
    }

    /** Rango de fechas inclusivo. */
    public record Rango(String fechaInicio, String fechaFin) {
    }

    /** Fecha de hoy en la zona horaria del negocio (por defecto Lima). */
    public static String hoy(String zonaHoraria) {
        return LocalDate.now(ZoneId.of(zonaHoraria)).toString();
    }

    /** true si el texto es una fecha real con formato AAAA-MM-DD. */
    public static boolean esFechaValida(String texto) {
        if (texto == null || !FORMATO.matcher(texto).matches()) return false;
        try {
            LocalDate.parse(texto); // ISO estricto: rechaza 2026-02-30
            return true;
        } catch (DateTimeParseException e) {
            return false;
        }
    }

    /** Dos rangos inclusivos que comparten al menos un día. */
    public static boolean seSuperponen(Rango a, Rango b) {
        return a.fechaInicio().compareTo(b.fechaFin()) <= 0 && b.fechaInicio().compareTo(a.fechaFin()) <= 0;
    }

    /** Devuelve el primer par de rangos que se superponen, o null. */
    public static Rango[] buscarSuperposicion(List<Rango> rangos) {
        List<Rango> ordenados = new ArrayList<>(rangos);
        ordenados.sort(Comparator.comparing(Rango::fechaInicio));
        for (int i = 1; i < ordenados.size(); i++) {
            if (seSuperponen(ordenados.get(i - 1), ordenados.get(i))) {
                return new Rango[]{ordenados.get(i - 1), ordenados.get(i)};
            }
        }
        return null;
    }

    /** Días de un rango inclusivo: del 01 al 01 = 1 día. */
    public static long diasEntre(String inicio, String fin) {
        return ChronoUnit.DAYS.between(LocalDate.parse(inicio), LocalDate.parse(fin)) + 1;
    }

    public static String sumarDias(String fecha, long dias) {
        return LocalDate.parse(fecha).plusDays(dias).toString();
    }
}
